src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/PreprocessingSymbolReferenceFinder.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.FindSymbols.Finders;
+
+internal class PreprocessingSymbolReferenceFinder : AbstractReferenceFinder<IPreprocessingSymbol>
+{
+    protected sealed override async ValueTask<ImmutableArray<FinderLocation>> FindReferencesInDocumentAsync(
+        IPreprocessingSymbol symbol,
+        FindReferencesDocumentState state,
+        FindReferencesSearchOptions options,
+        CancellationToken cancellationToken)
+    {
+        var tokens = await FindMatchingIdentifierTokensAsync(state, symbol.Name, cancellationToken).ConfigureAwait(false);
+
+        var normalReferences = await FindPreprocessingReferencesInTokensAsync(
+            symbol, state,
+            tokens.WhereAsArray(MatchesPreprocessingReference, state),
+            cancellationToken).ConfigureAwait(false);
+
+        return normalReferences;
+
+        static bool MatchesPreprocessingReference(SyntaxToken token, FindReferencesDocumentState state)
+        {
+            var syntaxFacts = state.SyntaxFacts;
+
+            var tokenParent = token.Parent;
+            Debug.Assert(tokenParent is not null);
+
+            // Quickly evaluate the common case that the parent is a #define or #undef directive
+            var parentKind = tokenParent.RawKind;
+            if (parentKind == syntaxFacts.SyntaxKinds.DefineDirectiveTrivia)
+                return true;
+
+            if (parentKind == syntaxFacts.SyntaxKinds.UndefDirectiveTrivia)
+                return true;
+
+            // In VB, a #Const directive assigns a value that can also
+            // derive from preprocessing symbols
```

Comment:
can you given an example?

Reply:
[Here](https://sharplab.io/#v2:EYLgtghglgdgNAGxAN2HALigpgJygMwE8AfAWACgBiAYQHsYBndAAgAUAlAUQ4HlrOAygICSAOQDiAfQEBNALIAhHgBlmAXmYAVdgFVOzA8wo16TZj00AJTu3VsuvfkLFTZilRWPD8zACKcFHXEtAAssGAoAcmYBWjAsZgBjWgATLGNOBAYsb2ZRC0kObnY+QREJaXklVQ0ixzKXSvdVTTCI8mjY+OZadDCcJNT0qk4YFOZvCkMjKlyLa1tW8IpRCHiGAAcIRISAMSnDagQIBgYkg+mDDYBXYAQoROYGW+Y5aBgACgBKC8vw8YEt1+/2YRxOZwoINW6y2OwyWWG0Kwm22CXEFAA5hAcMAIBiErjxgxCDB0BAAB5PF5vWDfX7TEGA4DAsag46nGZQtbI2HDSijca5IA==) is an example showing that capability. I was equally surprised to see it works, and that's why I thought it had to be supported.

Reply:
i'm not sure why this needs any special casing?  Isn't this just a matter of finding the matching names under any particular PP directive?

Reply:
In such a scenario, the parent of the node is not the directive itself, but the assignment operation, or even deeper for deeper expressions. This is specifically handled as part of the capabilities enabled by the expressions that are legal. I thought this could slightly improve performance on C# compilations that never need to go any further in the search.